DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs

```diff
@@ -232,10 +238,155 @@ public IActionResult ConfirmDeleteGroup(int groupId, ConfirmDeleteGroupViewModel
             return RedirectToAction("Index");
         }
 
+        [HttpGet]
+        [Route("Generate")]
+        public IActionResult GenerateGroups()
+        {
+            var registrationFieldOptions = GetRegistrationFieldOptionsSelectList();
+
+            var model = new GenerateGroupsViewModel(registrationFieldOptions);
+            return View(model);
+        }
+
+        [HttpPost]
+        [Route("Generate")]
+        public IActionResult GenerateGroups(GenerateGroupsViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                model.RegistrationFieldOptions = GetRegistrationFieldOptionsSelectList(model.RegistrationFieldOptionId);
+                return View(model);
+            }
+
+            var adminId = User.GetAdminId();
+            var centreId = User.GetCentreId();
+            var isJobGroup = model.RegistrationFieldOptionId == 7;
+            var linkedToField = GetLinkedToFieldValue((int)model.RegistrationFieldOptionId!);
+
+            var (newGroupNames, groupNamePrefix) = GetNewGroupNamesAndPrefix(model, isJobGroup);
+
+            using var transaction = new TransactionScope();
+            foreach (var (newGroupId, newGroupName) in newGroupNames)
+            {
+                var groupName = model.PrefixGroupName ? $"{groupNamePrefix} - {newGroupName}" : newGroupName;
+
+                var groupId = CreateNewGroup(model, adminId, centreId, linkedToField, groupName);
+
+                if (model.AddExistingDelegates && groupId != null)
+                {
+                    groupsService.AddDelegatesWithMatchingAnswersToGroup(
+                        (int)groupId,
+                        linkedToField,
+                        centreId,
+                        isJobGroup ? null : newGroupName,
+                        isJobGroup ? newGroupId : (int?)null
+                    );
+                }
+            }
+
+            transaction.Complete();
+
+            return RedirectToAction("Index");
+        }
+
         private IEnumerable<CustomPrompt> GetRegistrationPromptsWithSetOptions(int centreId)
         {
             return centreCustomPromptsService.GetCustomPromptsForCentreByCentreId(centreId).CustomPrompts
                 .Where(cp => cp.Options.Any());
         }
+
+        private IEnumerable<SelectListItem> GetRegistrationFieldOptionsSelectList(int? selectedId = null)
+        {
+            var registrationFieldOptions = GetCustomPromptOptions();
+
+            var jobGroupOption = (7, "Job group");
+            registrationFieldOptions.Add(jobGroupOption);

```

Comment:
I'm tempted to say that we should be trying to line these up with the LinkedToField values, since it would likely simplify some of the logic we need to do to convert between them.

Reply:
Ah I thought the linked to field value of 4 for Job groups was a foreign key for some reason, but it's not. I've changed the SQL in the data service method slightly from what was given in the ticket description so the RegistrationFieldOptionId will always align with the LinkedToField value.

Reply:
Linking the other relevant thread from the re-review here as well https://github.com/TechnologyEnhancedLearning/DLSV2/pull/930#discussion_r808158345

Reply:
Nevermind didn't realise it was important that the LinkedToField value 4 represents job group, changed back

Reply:
The reason for the mapping is I need to have the custom prompt number in the service method to get the custom prompt (GroupsService private method GetCustomPromptsAndPrefix). Having Job Group as 7 and then mapping in the service once to get the linked to field value seemed like the least about of conversion

Reply:
The new enumeration class RegistrationField now makes it easy to match the registration field id to the linked to field value